### PR TITLE
Call Webdrivers::Chromedriver.update in application_system_test_case.rb

### PR DESCRIPTION
**Pull request by rmacklin**
### Summary

Addresses #36288

This updates the templates for `application_system_test_case.rb` to explicitly trigger a chromedriver update. This is useful when system tests are run in parallel (the default mode for new Rails 6 applications), because it ensures that we only attempt to download chromedriver once before parallel processes are forked (rather than once per parallel process).

This should avoid the following error:
```
ChildProcess::LaunchError: Text file busy - /home/circleci/.webdrivers/chromedriver
```
which can happen if multiple parallel processes each try to download chromedriver.

### CODE REVIEW

File: railties/lib/rails/generators/rails/app/templates/test/application_system_test_case.rb.tt [hunk 1]
```diff
@@ -1,5 +1,7 @@
 require "test_helper"
 
+Webdrivers::Chromedriver.update
```
Comment (eileencodes) [inline railties/lib/rails/generators/rails/app/templates/test/application_system_test_case.rb.tt]: Sorry I misunderstood this change. Can this be added to Rails instead? I don't think it should be generated with the application. I think it's better for `driven_by` to call this.

File: railties/CHANGELOG.md [hunk 2]
```diff
@@ -1,5 +1,10 @@
 ## (Unreleased)
 
+*   Make generated application_system_test_case.rb files avoid
+    `ChildProcess::LaunchError` when parallel tests are enabled.
+
+    *Richard Macklin*
```
Comment (eileencodes) [inline railties/CHANGELOG.md]: Since this is a bug fix we don't need a changelog entry.

**Review by eileencodes — CHANGES_REQUESTED** [review]

File: railties/lib/rails/generators/rails/app/templates/test/application_system_test_case.rb.tt (hunk 1 above)
Comment (rmacklin) [reply railties/lib/rails/generators/rails/app/templates/test/application_system_test_case.rb.tt]: I'm looking at `driven_by`:
https://github.com/rails/rails/blob/5686c11c356223e9ca34093e88cdc9a616445355/actionpack/lib/action_dispatch/system_test_case.rb#L155-L159

which creates a [`SystemTesting::Driver`](https://github.com/rails/rails/blob/5686c11c356223e9ca34093e88cdc9a616445355/actionpack/lib/action_dispatch/system_testing/driver.rb) instance. Since the `webdrivers` gem is specific to `selenium`, I'm thinking `register_selenium` might be the right place. Does that seem right to you?

If so, perhaps something along these lines:
```diff
diff --git a/actionpack/lib/action_dispatch/system_testing/driver.rb b/actionpack/lib/action_dispatch/system_testing/driver.rb
index 25a09dd918..a90087f947 100644
--- a/actionpack/lib/action_dispatch/system_testing/driver.rb
+++ b/actionpack/lib/action_dispatch/system_testing/driver.rb
@@ -46,6 +46,8 @@ def register_selenium(app)
           Capybara::Selenium::Driver.new(app, { browser: @browser.type }.merge(browser_options)).tap do |driver|
             driver.browser.manage.window.size = Selenium::WebDriver::Dimension.new(*@screen_size)
           end
+
+          update_webdriver if Object.const_defined?(:Webdrivers)
         end
 
         def register_poltergeist(app)
@@ -61,6 +63,15 @@ def register_webkit(app)
         def setup
           Capybara.current_driver = @name
         end
+
+        def update_webdriver
+          case browser_options[:using]
+          when :chrome, :headless_chrome
+            Webdrivers::Chromedriver.update
+          when :firefox, :headless_firefox
+            Webdrivers::Geckodriver.update
+          end
+        end
     end
   end
 end
```

(The `if Object.const_defined?(:Webdrivers)` is because I'm assuming we want to support users who remove the `webdrivers` gem and opt to manage downloading/updating chromedriver themselves. I think maybe this should be its own configuration attribute that anyone could set, and `Object.const_defined?(:Webdrivers)` would be the default value. I'm not sure where that configuration option should go, or what we'd want to call it.)

Let me know if this is on the right track or if you had something different in mind.

File: railties/CHANGELOG.md (hunk 2 above)
Comment (rmacklin) [reply railties/CHANGELOG.md]: Sorry about that; I'll remove this with the other changes.

File: railties/lib/rails/generators/rails/app/templates/test/application_system_test_case.rb.tt (hunk 1 above)
Comment (rmacklin) [reply railties/lib/rails/generators/rails/app/templates/test/application_system_test_case.rb.tt]: @eileencodes sorry for letting this slip - I was hoping to get your thoughts on the above before proceeding (I don't want to go down the wrong path again if that's not what you were envisioning)

Comment (eileencodes) [reply railties/lib/rails/generators/rails/app/templates/test/application_system_test_case.rb.tt]: Hey sorry I've been really busy. How often does this actually need to be updated? This would get called always but does it need to get updated on every CI run?

Comment (rmacklin) [reply railties/lib/rails/generators/rails/app/templates/test/application_system_test_case.rb.tt]: Forgive me if I'm not understanding you, but I believe this already gets called whenever tests are run, due to these lines:
https://github.com/titusfortner/webdrivers/blame/v3.9.0/lib/webdrivers/chromedriver.rb#L155
https://github.com/titusfortner/webdrivers/blame/v3.9.0/lib/webdrivers/chromedriver.rb#L162

However, `Webdrivers::Chromedriver.update` will return early if chromedriver has already been updated:
https://github.com/titusfortner/webdrivers/blob/v3.9.0/lib/webdrivers/common.rb#L60

So with ephemeral machines as in the CircleCI example, each run will do one chromedriver update. The issue described in https://github.com/rails/rails/issues/36288 is because `driver_path` gets called lazily, which in practice is after parallel test processes have been forked, so we end up running into the
```
ChildProcess::LaunchError: Text file busy - /home/circleci/.webdrivers/chromedriver
```
exception.

In other words, I don't think calling `Webdrivers::Chromedriver.update` in `driven_by` would make it such that we're updating chromedriver any more often than before.

**Comment by rmacklin** [comment]
Closing in favor of #36592
